### PR TITLE
[CI] pytest-split uses `LeastDurationAlgorithm`

### PR DESCRIPTION
**Pull request by JerryChen97**
## Context

Two recurring issues with our CI test-splitting setup:

1. **Shard imbalance**: Core-test runner (4) consistently takes 2x+ longer than other runners (~23 min vs ~10 min). Root cause: `DurationBasedChunksAlgorithm` assigns tests in contiguous collection-order chunks, concentrating alphabetically-grouped heavy tests (e.g., FTQC MCM tests) into a single shard.

2. **Undetected duration anomalies**: A recent [auto-duration update](https://github.com/PennyLaneAI/pennylane/pull/9497) introduced severely inflated values (e.g., op0: 13s → 627s due to CI contention). With 50k+ line changes in the durations JSON, this is impossible to catch by human review. The inflated values further worsened shard imbalance, pushing core-tests (4) from ~23 min to ~57 min.

## Changes

1. **Switch to `least_duration` splitting algorithm** ([`LeastDurationAlgorithm`](https://jerry-git.github.io/pytest-split/api_docs/#pytest_split.algorithms.LeastDurationAlgorithm)) for both core-tests (6 shards) and jax-tests (4 shards). This is a greedy heap-based algorithm that assigns the heaviest unassigned test to the lightest shard, producing much better balance than the default contiguous-chunk approach.

2. **Add anomaly detection to the `update-durations` workflow.** Before creating the update PR, the workflow now compares new durations against the committed baseline. If any test shows a >10x duration increase (with old >1s and new >10s), the workflow fails with a detailed report and no PR is created.

3. **Fix pre-existing test ordering dependency** exposed by the new splitting: `test_postselect_mode_propagates_to_execute` added `"MidMeasure"` instead of `"MidMeasureMP"` to the device operations, making it silently depend on another test running first.

5. **Jax x64 preserver fixutre `preserve_jax_x64`**. Some tests use x32 jax but never recover the original `jax_enable_x64`. This issue was exposed by [one of the tests](https://github.com/PennyLaneAI/pennylane/actions/runs/26661096389/job/78583646305) after we re-shuffled the tests allocation in this PR.

## Benefits

1. **Better shard balance.** Observed improvement on this PR's CI (core tests):

   | Shard | main (`duration_based_chunks`) | PR (`least_duration`) |
   |-------|------|------|
   | 1 | 11.5 min | 16.4 min |
   | 2 | 10.1 min | 11.8 min |
   | 3 | 9.9 min | 15.9 min |
   | 4 | **22.8 min** | 17.5 min |
   | 5 | 11.4 min | 12.3 min |
   | 6 | 12.4 min | **19.2 min** |
   | **Max** | **22.8 min** | **19.2 min** |
   | **Spread** | 9.9–22.8 (2.3x) | 11.8–19.2 (1.6x) |

image

   With accurate durations, `least_duration` will maintain balance around near-optimal (bounded spread). In contrast, the previous `duration_based_chunks` algorithm has fundamentally different asymptotic behavior: as test suites grow or tests cluster alphabetically, it concentrates heavy tests into single shards with unbounded imbalance.

   **CI run links:**
   - main: https://github.com/PennyLaneAI/pennylane/actions/runs/26255763981 (commit `31500f3398`)
   - PR: https://github.com/PennyLaneAI/pennylane/actions/runs/26258382189 (commit `f657ae38e2`)

2. **Protection against catastrophic duration anomalies.** The update workflow will no longer silently commit contention-inflated values that cascade into shard imbalance regressions.

## Possible Drawbacks

1. **Current durations are still inaccurate** for some tests (particularly FTQC MCM tests due to a pre-existing Lightning slowdown on CI). Balance will improve incrementally as the durations file gets updated with accurate values.

2. **The anomaly detection workflow is not independently testable** in CI — it only triggers during the weekly `update-durations` run. However, the logic was validated against the PR #9497 data (correctly detects all 24 anomalous entries).

_Many thanks to **Copilot** for drafting the script, putting together words, and organizing paragraphs!_

**Related GitHub Issues:**
[sc-121180]

### CODE REVIEW

File: .github/workflows/update-durations.yml [hunk 1]
```diff
@@ -44,6 +44,76 @@ jobs:
...
+                  f.write(f"has_anomalies=true\n")
+                  # Write body to a file to avoid escaping issues
+              with open("anomaly_report.md", "w") as f:
+                  f.write(body.replace("\\n", "\n"))
+          else:
+              with open(os.environ["GITHUB_OUTPUT"], "a") as f:
+                  f.write("has_anomalies=false\n")
+              with open("anomaly_report.md", "w") as f:
+                  f.write("No duration anomalies detected.")
+          PYEOF
```
Comment (mudit2812) [inline .github/workflows/update-durations.yml]: This whole thing should be a Python script in the `.github/scripts` folder

Comment (JerryChen97) [reply .github/workflows/update-durations.yml]: https://github.com/PennyLaneAI/pennylane/pull/9519/commits/fcd26c0fc35483fbf5f2796bfb47f9345e6e3421

File: tests/conftest.py [hunk 2]
```diff
@@ -249,6 +249,21 @@ def pytest_generate_tests(metafunc):
...
+
+    Use this fixture on any test that sets ``jax_enable_x64`` to ``False`` so
+    that subsequent tests in the same xdist worker are not contaminated.
+    """
+    if not jax_available:
+        yield
+        return
+    original = jax.config.jax_enable_x64
+    yield
+    jax.config.update("jax_enable_x64", original)
```
Comment (astralcai) [inline tests/conftest.py:264]: Nice

Comment (astralcai) [reply tests/conftest.py:264]: Is the global default having x64 enabled? If so isn't it easier to have an autouse fixture that sets it to `True` before every test?

Comment (JerryChen97) [reply tests/conftest.py:264]: But jax tests are a small fraction, wouldn't it be a potential risk if we make the global fixure for all tests?

Comment (astralcai) [reply tests/conftest.py:264]: Good point...

File: .github/scripts/detect_duration_anomalies.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import argparse
+import json
+import os
+import sys
+
+# Thresholds
+MIN_OLD_DURATION = 1.0  # seconds – ignore short-running tests
+MIN_NEW_DURATION = 10.0  # seconds – only flag if the new value is "big"
+RATIO_THRESHOLD = 10  # flag if new/old exceeds this factor
+MAX_DISPLAY = 20  # cap table rows per file
```
Comment (astralcai) [inline .github/scripts/detect_duration_anomalies.py]: These values feel arbitrary. Is there a reason for those?

**Comment by mudit2812** [comment]
@JerryChen97 the spread has improved, but the total runtime of the tests has indeed gone up. There is a ~15 min increase in total runtime based on the table in the PR description

**Comment by JerryChen97** [comment]
> @JerryChen97 the spread has improved, but the total runtime of the tests has indeed gone up. There is a ~15 min increase in total runtime based on the table in the PR description

That is true. I think we need to pick between these two. For now I don't see other algorithms provided by pytest-split.
Originally I was more concerned about the currently in-used algorithm has been constantly accumulating more and more in the most expensive runner (core-4).

**Review by mudit2812 — COMMENTED** [review]
Thanks @JerryChen97 . The changes look fine to me. I'm just still a little weary about changing the splitting algorithm because of the increased runtime

**Comment by astralcai** [comment]
> Thanks @JerryChen97 . The changes look fine to me. I'm just still a little weary about changing the splitting algorithm because of the increased runtime

I'm wondering why would the total runtime change if we're simply splitting them differently across runners

File: .github/scripts/detect_duration_anomalies.py (hunk 3 above)
Comment (JerryChen97) [reply .github/scripts/detect_duration_anomalies.py]: I think actually the first two are not necessary.
The ratio threshold is indeed quite arbitrary. This is only from the fact that last several weeks there's a broken updating where there are cases the new duration is more than 10 times than before. So I wonder at least we won't allow such duration update to happen so just randomly set up a `10` here.

File: .github/scripts/detect_duration_anomalies.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+"""
+
+import argparse
+import json
+import os
+import sys
+
+# Thresholds
+MIN_OLD_DURATION = 1.0  # seconds – ignore short-running tests
+MIN_NEW_DURATION = 10.0  # seconds – only flag if the new value is "big"
```
Comment (JerryChen97) [inline .github/scripts/detect_duration_anomalies.py]:
```suggestion
```

File: .github/scripts/detect_duration_anomalies.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+def find_anomalies(old_data, new_data):
+    """Return a list of (test, old_dur, new_dur, ratio) tuples for anomalous tests."""
+    inflated = []
+    for test, new_dur in new_data.items():
+        old_dur = old_data.get(test, 0)
+        if old_dur > MIN_OLD_DURATION and new_dur > MIN_NEW_DURATION:
+            ratio = new_dur / old_dur
+            if ratio > RATIO_THRESHOLD:
+                inflated.append((test, old_dur, new_dur, ratio))
```
Comment (JerryChen97) [inline .github/scripts/detect_duration_anomalies.py]:
```suggestion
        ratio = new_dur / old_dur
        if ratio > RATIO_THRESHOLD:
            inflated.append((test, old_dur, new_dur, ratio))
```

File: .github/scripts/detect_duration_anomalies.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+def build_report(anomalies_by_file):
+    """Build a Markdown report string from anomaly results."""
+    lines = []
+    for fname, inflated in anomalies_by_file.items():
+        if not inflated:
+            continue
+        lines.append(f"### {fname}")
+        lines.append(
+            f"**{len(inflated)} tests with >{RATIO_THRESHOLD}x duration increase "
+            f"(old >{MIN_OLD_DURATION}s, new >{MIN_NEW_DURATION}s):**"
```
Comment (JerryChen97) [inline .github/scripts/detect_duration_anomalies.py]:
```suggestion
```

File: .github/scripts/detect_duration_anomalies.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        --report anomaly_report.md
+"""
+
+import argparse
+import json
+import os
+import sys
+
+# Thresholds
+RATIO_THRESHOLD = 10  # flag if new/old exceeds this factor
```
Comment (astralcai) [inline .github/scripts/detect_duration_anomalies.py]: I feel that 10 sounds a bit high. Imagine a test that used to take a minute, now it takes 9 minutes, and it wouldn't be flagged by the system?

Comment (JerryChen97) [reply .github/scripts/detect_duration_anomalies.py]: I agree. How about 2, or even lower?

File: .github/scripts/detect_duration_anomalies.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        lines.append("|------|-----|-----|-------|")
+        for test, od, nd, r in inflated[:MAX_DISPLAY]:
+            short = test.split("::")[-1][:60]
+            lines.append(f"| `{short}` | {od:.1f}s | {nd:.1f}s | **{r:.0f}x** |")
+        if len(inflated) > MAX_DISPLAY:
+            lines.append(f"| ... and {len(inflated) - MAX_DISPLAY} more | | | |")
+
+    if lines:
+        header = (
+            "## \u26a0\ufe0f Duration Anomalies Detected\n\n"
```
Comment (astralcai) [inline .github/scripts/detect_duration_anomalies.py]: I don't think we need emojis in the printout lol. This gives off strong AI vibes 😆

Comment (mudit2812) [reply .github/scripts/detect_duration_anomalies.py]: GH copilot shenanigans I guess

**Comment by mudit2812** [comment]
> > Thanks @JerryChen97 . The changes look fine to me. I'm just still a little weary about changing the splitting algorithm because of the increased runtime
> 
> I'm wondering why would the total runtime change if we're simply splitting them differently across runners

@astralcai my guess is that there is some additional state-keeping necessary, causing the slowdown

File: .github/scripts/detect_duration_anomalies.py (hunk 8 above)
Comment (JerryChen97) [reply .github/scripts/detect_duration_anomalies.py]: Yeah let me cherry pick out those Claude favorite emojis :sweat:

**Comment by JerryChen97** [comment]
> > > Thanks @JerryChen97 . The changes look fine to me. I'm just still a little weary about changing the splitting algorithm because of the increased runtime
> > 
> > 
> > I'm wondering why would the total runtime change if we're simply splitting them differently across runners
> 
> @astralcai my guess is that there is some additional state-keeping necessary, causing the slowdown

From my observation, the two algorithms provided by pytest-split: `duration_based_chunks` (which we used to select by default) keeps the original order of tests and did minimal split, so it made those tests that belong to the same context run together; meanwhile `least-duration` only cares about duration, and it bin-packs tests in a greedy way, leading to fragmentation, so eventually many tests end up re-loading shared modules (or even re-qjit I guess?) hence naturally bringing slow-down.

An interesting side-effect is that such fragmentation revealed certain amount of test issues e.g. secret dependency between different unit tests and those forgotten reset of jax x64, which used to be covered by the contiguous distribution defined by `duration_based_chunks`

File: .github/scripts/detect_duration_anomalies.py (hunk 8 above)
Comment (JerryChen97) [reply .github/scripts/detect_duration_anomalies.py]: https://github.com/PennyLaneAI/pennylane/pull/9519/commits/0817c5b5c58380d8720b94ed4cccf0d45d5e2118

File: .github/scripts/detect_duration_anomalies.py (hunk 7 above)
Comment (JerryChen97) [reply .github/scripts/detect_duration_anomalies.py]: https://github.com/PennyLaneAI/pennylane/pull/9519/commits/0817c5b5c58380d8720b94ed4cccf0d45d5e2118

Comment (astralcai) [reply .github/scripts/detect_duration_anomalies.py]: I'd say 3, and maybe we should bring back the lower bound. Imagine a test that used to take 0.05 seconds now taking 0.1 seconds. It's probably not a big deal. For shorter tests where the overhead of whatever might dominate the actual test runtime, the ratio is not really a good indicator.

Comment (JerryChen97) [reply .github/scripts/detect_duration_anomalies.py]: Yeah that's a good idea. I think even further I need to bring some duration statistics here so that we have a better idea what kinda distribution we are facing in our CI workflows.

**Comment by JerryChen97** [comment]
A quick re-cap:
I don't think I ever see a good enough solution here yet: either algorithm has almost unacceptable drawbacks; and I am not aware a third algorithm yet inside [pytest-split](https://github.com/jerry-git/pytest-split/blob/master/src/pytest_split/algorithms.py). This smells more of an upstream issue.
Since it's not yet something that is gonna destroy our life immediately, I would say let's be conservative given the current method is running well (core part 4 is still accumulating at runtime, but tolerable).

However, other things that were found along the way here could be nicely split into independent PR's:
1. Decoupling of the tests that were found to be inter-dependent on each other
2. The duration update guard script

File: tests/devices/test_default_qutrit.py [hunk 9]
```diff
@@ -1543,6 +1543,7 @@ class TestDtypePreservedJax:
             qp.probs(wires=[2, 0]),
         ],
     )
+    @pytest.mark.usefixtures("preserve_jax_x64")
```
Comment (JerryChen97) [inline tests/devices/test_default_qutrit.py]:
```suggestion
```

File: tests/devices/test_default_qutrit.py [hunk 10]
```diff
@@ -1566,6 +1567,7 @@ def circuit(x):
 
     @pytest.mark.usefixtures("preserve_jax_x64")
     @pytest.mark.parametrize("enable_x64, c_dtype", [(False, np.complex64), (True, np.complex128)])
+    @pytest.mark.usefixtures("preserve_jax_x64")
```
Comment (JerryChen97) [inline tests/devices/test_default_qutrit.py]:
```suggestion
```

File: tests/workflow/interfaces/qnode/test_jax_jit_qnode.py [hunk 11]
```diff
@@ -3366,6 +3366,7 @@ def test_type_conversion_fallback(self):
 
     @pytest.mark.usefixtures("preserve_jax_x64")
     @pytest.mark.parametrize("diff_method", ("adjoint", "parameter-shift"))
+    @pytest.mark.usefixtures("preserve_jax_x64")
```
Comment (JerryChen97) [inline tests/workflow/interfaces/qnode/test_jax_jit_qnode.py]:
```suggestion
```

File: tests/workflow/interfaces/qnode/test_jax_jit_qnode.py [hunk 12]
```diff
@@ -3381,6 +3382,7 @@ def circuit(x):
 
     @pytest.mark.usefixtures("preserve_jax_x64")
     @pytest.mark.parametrize("diff_method", ("adjoint", "finite-diff"))
+    @pytest.mark.usefixtures("preserve_jax_x64")
```
Comment (JerryChen97) [inline tests/workflow/interfaces/qnode/test_jax_jit_qnode.py]:
```suggestion
```

File: tests/test_compiler.py [hunk 13]
```diff
@@ -108,6 +108,7 @@ def circuit(phi, theta):
 
     @pytest.mark.usefixtures("preserve_jax_x64")
     @pytest.mark.parametrize("jax_enable_x64", [False, True])
+    @pytest.mark.usefixtures("preserve_jax_x64")
```
Comment (JerryChen97) [inline tests/test_compiler.py]:
```suggestion
```

File: .github/scripts/detect_duration_anomalies.py (hunk 7 above)
Comment (JerryChen97) [reply .github/scripts/detect_duration_anomalies.py]: Did both https://github.com/PennyLaneAI/pennylane/pull/9519/commits/ae7d7bc830925ae354ea05c07e591014201bbf8b @astralcai

**Review by mudit2812 — COMMENTED** [review]
Have you had a chance to try out the script you added in a workflow? If not, I would prefer if we just merge the change to the splitting algorithm for now and add the anomaly detection in a follow up PR

**Comment by JerryChen97** [comment]
> Have you had a chance to try out the script you added in a workflow? If not, I would prefer if we just merge the change to the splitting algorithm for now and add the anomaly detection in a follow up PR

I think I better just split them up anyways. I feel genuinely they should just be in different PR

**Comment by JerryChen97** [comment]
Now the algorithm switching and the anomaly detection have been separated! The other PR is https://github.com/PennyLaneAI/pennylane/pull/9664
@mudit2812 @astralcai

**Review by mudit2812 — APPROVED** [review]

**Review by astralcai — APPROVED** [review]
